### PR TITLE
Allow constants to refer statics in const eval

### PR DESCRIPTION
**Pull request by rpjohnst**
This is the dynamic portion of what we discussed in rust-lang/const-eval#11.
Changes to the static check are left to a separate PR.

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/borrow_check/type_check/mod.rs [hunk 1]
```diff
@@ -366,20 +366,6 @@ impl<'a, 'b, 'tcx> Visitor<'tcx> for TypeVerifier<'a, 'b, 'tcx> {
                         );
                     }
                 }
-            } else if let Some(static_def_id) = constant.check_static_ptr(tcx) {
```
Comment (RalfJung) [inline src/librustc_mir/borrow_check/type_check/mod.rs:369]: What is this diff doing here? This is changing the *borrow checker*, which should be unnecessary when all you want to do is change the dynamic const-to-static check.

File: src/librustc_mir/const_eval/eval_queries.rs [hunk 2]
```diff
@@ -94,10 +94,13 @@ pub(super) fn mk_eval_cx<'mir, 'tcx>(
     )
 }
 
-pub(super) fn op_to_const<'tcx>(
+#[derive(Debug)]
+pub(super) struct RefersToStatic;
```
Comment (RalfJung) [inline src/librustc_mir/const_eval/eval_queries.rs:98]: I don't think we need a new error type here. There are tons of errors that can happen in const-eval, why is this one so special that it has to shop up here? There is no precedent for anything like this.

File: src/librustc_mir/const_eval/machine.rs [hunk 3]
```diff
@@ -368,16 +366,14 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter {
             if memory_extra.can_access_statics {
```
Comment (RalfJung) [inline src/librustc_mir/const_eval/machine.rs:366]: This should be rename to `can_access_mutable_globals` or so.

File: src/librustc_mir/const_eval/machine.rs [hunk 4]
```diff
@@ -368,16 +366,14 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter {
...
                 // Machine configuration allows us read from anything (e.g., `static` initializer).
                 Ok(())
-            } else if static_def_id.is_some() {
-                // Machine configuration does not allow us to read statics
+            } else if allocation.mutability != Mutability::Not {
```
Comment (RalfJung) [inline src/librustc_mir/const_eval/machine.rs:369]:
```suggestion
            } else if allocation.mutability == Mutability::Mut {
```

File: src/librustc_mir/interpret/validity.rs [hunk 5]
```diff
@@ -411,11 +411,11 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, 'tcx, M
                     if !did.is_local() || self.ecx.tcx.is_foreign_item(did) {
                         return Ok(());
                     }
+                    // FIXME: Statics referenced from consts are skipped.
+                    // This avoids spurious "const accesses static" errors
+                    // unrelated to validity, but is similarly unsound.
```
Comment (RalfJung) [inline src/librustc_mir/interpret/validity.rs:416]: I'd merge this with the `if` above, to
```rust
if !did.is_local() || self.ecx.tcx.is_static(did) {
```
The comment should also be updated:
```
// We don't re-validate statics because 
// - the const-machine we are in might forbid reading from them, and
// - they have already been validated.
// FIXME: validation might have happened at a different type, but for now
// we consider this an okay trade-off.
```

Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs:416]: But your check is stricter in the sense that it validates more, so maybe @oli-obk would prefer that one.

I just think it's generally not worth recursing into other globals -- this can only possibly catch a bug if people did some crazy unsafe stuff, and those people asked for it when they get post-monomorphization errors. So I'd actually be fine to just always bail for any `Some(GlobalAlloc::Static(_))` here.

File: src/librustc_mir_build/build/expr/as_constant.rs [hunk 6]
```diff
@@ -21,7 +21,8 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         let Expr { ty, temp_lifetime: _, span, kind } = expr;
         match kind {
             ExprKind::Scope { region_scope: _, lint_level: _, value } => this.as_constant(value),
-            ExprKind::Literal { literal, user_ty } => {
+            ExprKind::Literal { literal, user_ty } |
+            ExprKind::StaticRef { literal, user_ty, .. } => {
```
Comment (RalfJung) [inline src/librustc_mir_build/build/expr/as_constant.rs:25]: Here and below we have more strange changes that seem entirely unrelated to the PR description.

File: src/test/ui/consts/const-points-to-static.rs [hunk 7]
```diff
@@ -1,10 +1,10 @@
+// check-pass
 // compile-flags: -Zunleash-the-miri-inside-of-you
```
Comment (RalfJung) [inline src/test/ui/consts/const-points-to-static.rs:2]: This test should be moved to the miri-unleashed folder.

Comment (RalfJung) [reply src/test/ui/consts/const-points-to-static.rs:2]: Also it should then be adjusted to check that reading the static actually works. Like,
```rust
const TEST_VAL: u8 = *TEST;

fn main() { assert_eq!(TEST_VAL, 4); }
```

File: src/test/ui/consts/const-prop-read-static-in-const.rs [hunk 8]
```diff
@@ -1,8 +1,9 @@
+// check-pass
 // compile-flags: -Zunleash-the-miri-inside-of-you
 
 #![allow(dead_code)]
 
-const TEST: u8 = MY_STATIC; //~ ERROR any use of this value will cause an error
```
Comment (RalfJung) [inline src/test/ui/consts/const-prop-read-static-in-const.rs:5]: This, too, should get an assertion in main checking the value that was read.
Also Cc @ecstatic-morse as this is a const-prop test; I am not entirely sure what it is supposed to test and why it unleashes Miri so it is unclear how to adjust it.

File: src/test/ui/consts/miri_unleashed/const_refers_to_static.stderr [hunk 9]
```diff
@@ -1,59 +1,73 @@
...
-LL | / const REF_INTERIOR_MUT: &usize = {
-LL | |     static FOO: AtomicUsize = AtomicUsize::new(0);
-LL | |     unsafe { &*(&FOO as *const _ as *const usize) }
-LL | |
-LL | | };
-   | |__^ type validation failed: encountered a reference pointing to a static variable
+LL | const DEREF_INTERIOR_MUT: usize = *REF_INTERIOR_MUT;
+   | ----------------------------------^^^^^^^^^^^^^^^^^-
+   |                                   |
+   |                                   constant accesses static
```
Comment (RalfJung) [inline src/test/ui/consts/miri_unleashed/const_refers_to_static.stderr:61]: The error message should be reworded now, something like "constant reading from mutable static".

File: src/test/ui/consts/miri_unleashed/const_refers_to_static.stderr [hunk 10]
```diff
@@ -82,25 +90,20 @@ LL | | };
...
-LL | | };
-   | |__^ type validation failed: encountered a reference pointing to a static variable
+warning: any use of this value will cause an error
+  --> $DIR/const_refers_to_static.rs:53:37
    |
-   = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rustc repository if you believe it should not be considered undefined behavior.
+LL | const READ_EXTERN: usize = unsafe { EXTERN };
+   | ------------------------------------^^^^^^---
+   |                                     |
+   |                                     cannot read from foreign (extern) static DefId(0:11 ~ const_refers_to_static[317d]::[0]::EXTERN[0])
```
Comment (RalfJung) [inline src/test/ui/consts/miri_unleashed/const_refers_to_static.stderr:106]: That error message doesn't look great.

File: src/test/ui/consts/miri_unleashed/use-static-via-const.rs [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    match p {
+        p if p == PTR => true,
+        _ => false,
+    }
+}
+
+fn main() {
+    assert!(parse_slice(&DATA));
+    assert!(parse_id(DATA.as_ptr()));
+}
```
Comment (RalfJung) [inline src/test/ui/consts/miri_unleashed/use-static-via-const.rs:29]: This doesn't actually *read* from the static in a const, does it?

Also I feel we are getting a lot of duplicate tests here, some of the ones above seem to already test the same thing?

File: src/librustc_mir/borrow_check/type_check/mod.rs (hunk 1 above)
Comment (rpjohnst) [reply src/librustc_mir/borrow_check/type_check/mod.rs:369]: I guess I should have been clearer in the commit message. This check assumes that the only constants that can point to statics are the ones created and immediately dereferenced [during MIR building](https://github.com/rust-lang/rust/pull/71332/files#diff-33647ed7aa1d34b0fdc2c36d843a8169), so it fails with anything else. (For example, a constant pointing to a static's field, which has a different type than the static itself.)

So the first commit of this PR replaces this check with a more precise one suggested (AFAIU) by @eddyb [when it was added](https://github.com/rust-lang/rust/pull/69145/files#r382089824). I guess this could also make sense as a separate PR? Or at least some borrowck-oriented review?

File: src/librustc_mir/interpret/validity.rs (hunk 5 above)
Comment (rpjohnst) [reply src/librustc_mir/interpret/validity.rs:416]: I went with the stricter check because there are tests for this, for statics-pointing-to-statics:
* https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const-eval/double_check.rs
* https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const-eval/double_check2.rs

Happy to collapse it into the first check if we're okay ditching those tests.

File: src/test/ui/consts/miri_unleashed/use-static-via-const.rs (hunk 11 above)
Comment (rpjohnst) [reply src/test/ui/consts/miri_unleashed/use-static-via-const.rs:29]: Hm, I added this test because it caught some ICEs that the others didn't- one related to the borrowck change and another related to the fallback-to-ByRef changes. Perhaps that code would also be exercised by some of your suggestions above? I'll look into it.

File: src/librustc_mir/borrow_check/type_check/mod.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/borrow_check/type_check/mod.rs:369]: So without this, the rest of the PR fails, or so?

I certainly cannot review anything in borrw_check or hair, I have never touched or read any of that code.^^

File: src/librustc_mir/interpret/validity.rs (hunk 5 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs:416]: Yeah I think it is okay to not catch the failure in double_check2. We already do not catch it cross-crate. But let's wait what @oli-obk thinks -- and in fact let's Cc @rust-lang/wg-const-eval and see what everyone else thinks :D

File: src/test/ui/consts/miri_unleashed/use-static-via-const.rs (hunk 11 above)
Comment (RalfJung) [reply src/test/ui/consts/miri_unleashed/use-static-via-const.rs:29]: If this caused extra ICEs, that seems fair then. Please add a comment in the test explaining that.

File: src/librustc_mir/borrow_check/type_check/mod.rs (hunk 1 above)
Comment (rpjohnst) [reply src/librustc_mir/borrow_check/type_check/mod.rs:369]: Yes, without this many scenarios using consts that refer to statics will fail at that `span_mirbug!`.

File: src/test/ui/consts/const-prop-read-static-in-const.rs (hunk 8 above)
Comment (ecstatic-morse) [reply src/test/ui/consts/const-prop-read-static-in-const.rs:5]: Why am I cc'ed?

Comment (RalfJung) [reply src/test/ui/consts/const-prop-read-static-in-const.rs:5]: Didn't you do const-prop stuff? Or am I mixing that up with someone else?

Comment (RalfJung) [reply src/test/ui/consts/const-prop-read-static-in-const.rs:5]: Oh I think I meant @wesleywiser, sorry for that!

Comment (wesleywiser) [reply src/test/ui/consts/const-prop-read-static-in-const.rs:5]: I think this was added just to document the current state when `-Zunleash-the-miri-inside-of-you` is enabled. This change seems fine to me. Asserting in main that the value was read like @RalfJung suggested also sounds like a good idea to me.

File: src/librustc_mir/interpret/validity.rs (hunk 5 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/validity.rs:416]: wait, how do we not catch this cross crate? Don't we revisit statics from other crates if we encounter them at a different type of non-zero offset?

File: src/librustc_mir/interpret/validity.rs [hunk 12]
```diff
@@ -411,11 +411,11 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, 'tcx, M
...
                     if !self.may_ref_to_static && self.ecx.tcx.is_static(did) {
-                        throw_validation_failure!(
```
Comment (oli-obk) [inline src/librustc_mir/interpret/validity.rs:415]: Can't we keep this validation failure in case the static's memory is mutable? or do we already report errors if constants refer to mutable memory (i kinda remember talking about this with @RalfJung on a recent PR)

File: src/librustc_mir/const_eval/eval_queries.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/const_eval/eval_queries.rs:98]: This is not actually an error. This is a signal that we should turn this into a `ByRef` constant instead of trying to use `Scalar` or `Slice`. At the very least it must be documented to explain this properly, but imo you an also just use `Option` as the return type of `try_op_to_const`

File: src/librustc_mir/const_eval/eval_queries.rs [hunk 13]
```diff
@@ -128,10 +131,16 @@ pub(super) fn op_to_const<'tcx>(
         op.try_as_mplace(ecx)
     };
 
+    let alloc_map = ecx.tcx.alloc_map.lock();
+    let try_unwrap_memory = |alloc_id: AllocId| match alloc_map.get(alloc_id) {
+        Some(GlobalAlloc::Memory(mem)) => Ok(mem),
+        Some(GlobalAlloc::Static(_)) => Err(RefersToStatic),
```
Comment (oli-obk) [inline src/librustc_mir/const_eval/eval_queries.rs:137]: This breaks our assumption, that any constant of scalar type or `&str`/`&[u8]` type is `ConstValue::Scalar` or `ConstValue::Slice` respectively. If such a constant is used in pattern matching, things will break in bad ways (e.g. two identical patterns not comparing equal). I guess such constants must be forbidden from occurring in patterns and const generics.

File: src/librustc_mir/interpret/validity.rs (hunk 5 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs:416]: No we don't, as far as I know.

File: src/librustc_mir/interpret/validity.rs (hunk 12 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs:415]: The thing is, from a dynamic perspective there is nothing wrong with a const that *points to* a mutable static. So I don't think we should check for it here.

The static checks do that as part of their approximation, sure, but quoting from https://github.com/rust-lang/rust/issues/49146#issuecomment-616301045
> Miri is perfectly capable of emitting an error when it encounters an illegal operation on a qualified value, and it can do so with no false positives. 

I agree with @ecstatic-morse that it *should* do so with no false positives. But failing with a const *refers to* (but does not read from) a mutable static is a false positive.

File: src/librustc_mir/const_eval/eval_queries.rs (hunk 13 above)
Comment (RalfJung) [reply src/librustc_mir/const_eval/eval_queries.rs:137]: Oh... pattern matching assumes that references in consts can only point to immutable memory? That's... a rather big assumption! It was true so far but only due to extreme conservatism on the side of our static const checks.

However, I feel like the fact that `&str` is an immutable shared reference should count for something. Maybe interning should ensure that shared references only point to read-only allocations, or so? (recursively)

Comment (oli-obk) [reply src/librustc_mir/const_eval/eval_queries.rs:137]: If the `&str` came from a `&mut str`, we can't turn the `&mut str` allocation into an immutable one. And if we allow constants to refer to statics, that can happen, and we must keep the pointer to the static, because otherwise we violate the fact that a pointer to the same static must have the same address.

Comment (RalfJung) [reply src/librustc_mir/const_eval/eval_queries.rs:137]: How'd that look like though? When a const takes an `&mut` to something, that reference has lifetime `'static`, so nothing else ever in the entire program is allowed to access the memory behind that unique reference.

Though when we have `&i32`, that could indeed be from an `&mut i32` that only covers part of an allocation and the rest of it could still permit mutation.

I think allowing references in consts in patterns was a big mistake... do we permit this for all types or just `&str`/`&[u8]`?

Comment (oli-obk) [reply src/librustc_mir/const_eval/eval_queries.rs:137]: > I think allowing references in consts in patterns was a big mistake... do we permit this for all types or just &str/&[u8]?

Hold up :D we need special rules for pattern constants and const generics anyway. @eddyb talked to me about this already. E.g. we can't non-normalized  padding or non-interned references in constants used in const generics. We do need a separate interning & normalization for such constants anyway. Not sure how we can detect references into statics though, maybe we need a new flag in `Alloation` that marks the allocation as part of a static, and thus having a unique address.

Comment (RalfJung) [reply src/librustc_mir/const_eval/eval_queries.rs:137]: I opened https://github.com/rust-lang/const-eval/issues/42 for the soundness concerns.

We probably have to get that answered before allowing consts pointing to mutable statics in our dynamic checks. Sorry @rpjohnst, I was entirely unaware of this particular minefield.

For now, could you remove those changes from this PR? I.e., dynamically permit consts pointing to *immutable* statics, but make no effort to permit consts pointing to *mutable* statics?

Comment (oli-obk) [reply src/librustc_mir/const_eval/eval_queries.rs:137]: As long as we do it behind a feature gate, that's perfectly fine in my book.

Comment (RalfJung) [reply src/librustc_mir/const_eval/eval_queries.rs:137]: @rpjohnst's use-case doesn't need consts pointing to mutable statics though. I think this change is part of this PR because I recommended them to do that, because I thought we could weaken our dynamic checks all the way to "just prevent reads from mutable statics, everything else is fine".

I was wrong as I didn't know about pattern-matching-soundness. In light of this new information, I'd prefer it we didn't weaken our dynamic checks quite so far.

Comment (rpjohnst) [reply src/librustc_mir/const_eval/eval_queries.rs:137]: Sure, will do. To clarify- do we want these dynamic checks behind a feature gate as well? Or can we still rely on the static check and just feature gate any changes there?

Comment (RalfJung) [reply src/librustc_mir/const_eval/eval_queries.rs:137]: If you avoid touching validity (beyond what is necessary to enable consts to point to immutable statics), I don't think any feature gates are needed for weakening the dynamic checks.

**Comment by crlf0710** [comment]
@rpjohnst Ping from triage, need a rebase here.

**Comment by rpjohnst** [comment]
This is blocked on deciding what to actually do, I can close it for now.
